src/path.cr

```diff
@@ -0,0 +1,942 @@
+  #
+  # ```
+  # Path.posix("foo/bar").join(Path.windows("baz\baq"))  # => Path.posix("foo/bar/baz/baq")
+  # Path.windows("foo\\bar").join(Path.posix("baz/baq")) # => Path.posix("foo\\bar\\baz/baq")
+  # ```
+  def join(parts : Enumerable) : Path
+    new_name = String.build do |str|
+      str << @name
+      last_ended_with_separator = ends_with_separator?
+
+      parts.each_with_index do |part, index|
+        case part
+        when Path
+          # Every POSIX path is also a valid Windows path, so we only need to
+          # convert the other way around (see `#to_windows`, `#to_posix`).
+          part = part.to_posix if posix? && part.windows?
+
+          part = part.@name
+        else
+          part = part.to_s
+          part.check_no_null_byte
+        end
+
+        if part.empty?
+          if index == parts.size - 1
+            str << separators[0] unless last_ended_with_separator
+            last_ended_with_separator = true
+          else
+            last_ended_with_separator = false
+          end
+
+          next
+        end
+
+        byte_start = 0
+        byte_count = part.bytesize
+
+        case {starts_with_separator?(part), last_ended_with_separator}
+        when {true, true}
+          byte_start += 1
+          byte_count -= 1
+        when {false, false}
+          str << separators[0] unless str.bytesize == 0
+        end
+
+        last_ended_with_separator = ends_with_separator?(part)
+
+        str.write part.unsafe_byte_slice(byte_start, byte_count)
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Appends the given *part* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"] / "bar" / "baz"     # => Path["foo/bar/baz"]
+  # Path["foo/"] / Path["/bar/baz"] # => Path["foo/bar/baz"]
+  # ```
+  def /(part : Path | String) : Path
+    join(part)
+  end
+
+  # Resolves path *name* in this path's parent directory.
+  #
+  # Raises `Path::Error` if `#parent` is `nil`.
+  def sibling(name : Path | String) : Path?
+    if parent = self.parent
+      parent.join(name)
+    else
+      raise Error.new("Can't resolve sibling for a path without parent directory")
+    end
+  end
+
+  # Compares this path to *other*.
+  #
+  # The comparison is performed strictly lexically: `foo` and `./foo` are *not*
+  # treated as equal. To compare paths semantically, they need to be normalized
+  # and converted to the same kind.
+  #
+  # ```
+  # Path["foo"] <=> Path["foo"]               # => 0
+  # Path["foo"] <=> Path["./foo"]             # => 1
+  # Path.posix("foo") <=> Path.windows("foo") # => -1
+  # ```
+  #
+  # Comparison is case-sensitive for POSIX paths and case-insensitive for
+  # Windows paths.
+  #
+  # ```
+  # Path.posix("foo") <=> Path.posix("FOO")     # => 1
+  # Path.windows("foo") <=> Path.windows("FOO") # => 0
+  # ```
+  def <=>(other : Path)
+    ord = @name.compare(other.@name, case_insensitive: windows?)
+    return ord if ord != 0
+
+    @kind <=> other.@kind
```

Comment:
I find this strange, but we can probably fix it later. Why a POSIX path is less than a WINDOWS path?

(and if that makes sense... shouldn't it be the other way around? 😛)

Reply:
Also the PartialComparable is no longer needed due to latest changes in Comparable. But that can be changed later.

Reply:
Hm, you are right! I can fix that in a PR after this one was merged (also removing `PartialComparable`).

Reply:
~~removing~~ deprecating 🙈

Reply:
Yes, though it isn't used anywhere else according to GitHub: https://github.com/search?q=language%3Acrystal+PartialComparable&type=Code

Reply:
Regarding the original question: Two paths shouldn't be treated as equal if they are of different kinds, because that means different semantics. They shouldn't compare equal either, so this seems like a legitimate solution.
Could alternatively return `nil` of course when comparing different path types.

Reply:
Returning nil sounds good to me